### PR TITLE
AMDGPU: Start selecting flat/global atomicrmw fmin/fmax.

### PR DESCRIPTION
**Pull request by arsenm**
Define subtarget features for atomic fmin/fmax support.

The flat/global support is a real messe. We had float/double support at
the beginning in gfx6 and gfx7. gfx8 removed these. gfx10 reintroduced them.
gfx11 removed the f64 versions again.

gfx9 partially reintroduced them, in gfx90a and gfx940 but only for f64.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95592?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#95592** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95592?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#95591** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95591?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95590** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/FLATInstructions.td [hunk 1]
```diff
@@ -1582,33 +1603,33 @@ let OtherPredicates = [isGFX12Plus] in {
...
-defm : GlobalFLATAtomicPats <"GLOBAL_ATOMIC_MIN_F64", "atomic_load_fmin_global", f64>;
-defm : GlobalFLATAtomicPats <"GLOBAL_ATOMIC_MAX_F64", "atomic_load_fmax_global", f64>;
-defm : GlobalFLATAtomicIntrPats <"GLOBAL_ATOMIC_MIN_F64", "int_amdgcn_global_atomic_fmin", f64>;
-defm : GlobalFLATAtomicIntrPats <"GLOBAL_ATOMIC_MAX_F64", "int_amdgcn_global_atomic_fmax", f64>;
-defm : FlatSignedAtomicPat <"FLAT_ATOMIC_MIN_F64", "atomic_load_fmin_flat", f64>;
-defm : FlatSignedAtomicPat <"FLAT_ATOMIC_MAX_F64", "atomic_load_fmax_flat", f64>;
-defm : FlatSignedAtomicIntrPat <"FLAT_ATOMIC_MIN_F64", "int_amdgcn_flat_atomic_fmin", f64>;
-defm : FlatSignedAtomicIntrPat <"FLAT_ATOMIC_MAX_F64", "int_amdgcn_flat_atomic_fmax", f64>;
-}
+// let OtherPredicates = [isGFX10Only] in { // fixme
```
Comment (krzysz00) [inline llvm/lib/Target/AMDGPU/FLATInstructions.td:1620]: Why commented out?

**Review by krzysz00 — COMMENTED** [review]
I'm not seeing anything obviously wrong here, but I don't know if I'm the right person to approve this in

File: llvm/lib/Target/AMDGPU/FLATInstructions.td (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/FLATInstructions.td:1620]: We had duplicated pseudos and handling on different sub targets, I unified these in #95591 so this should be removable

File: llvm/lib/Target/AMDGPU/BUFInstructions.td [hunk 2]
```diff
@@ -1699,7 +1709,7 @@ multiclass SIBufferAtomicPat_Common<string OpPrefix, ValueType vt, string Inst,
 
 multiclass SIBufferAtomicPat<string OpPrefix, ValueType vt, string Inst,
                              list<string> RtnModes = ["ret", "noret"]> {
-  let SubtargetPredicate = HasUnrestrictedSOffset in {
+  let OtherPredicates = [HasUnrestrictedSOffset] in {
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/BUFInstructions.td:1712]: A side question, what is the difference between `OtherPredicates` and `SubtargetPredicate`? It looks like you swapped a couple of them here.

**Review by shiltian — APPROVED** [review]
LG

File: llvm/lib/Target/AMDGPU/BUFInstructions.td (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/BUFInstructions.td:1712]: SubtargetPredicate is just one thing, and OtherPredicates is a list? The usage pattern consistency has been declining for a while

**Comment by arsenm** [comment]
### Merge activity

* **Jun 23, 4:06 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/95592).

**Pull request merged**